### PR TITLE
[New Scheduler] Add memory queue for the new scheduler

### PR DESCRIPTION
**Pull request by style95**
## Description
This PR adds one of the core implementations of the new scheduler.
This queue is dynamically created for each action and buffer activations in an internal queue.
It provisions more containers according to the number of activations in the queue.
This process is called "scheduling" and it is handled by another component `SchedulingDecisionMaker`.

The queue communicates with `ContainerManager` to add more containers.
`ContainerManager` selects proper invokers and sends container creation messages to them.

This queue is designed in a way that activations are not persistent.
But it could be improved in the future by taking a similar approach with what Kafka does.

## My changes affect the following components

- [ ] API
- [ ] Controller
- [x] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [x] Scheduler
- [x] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [x] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [x] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [x] My changes require further changes to the documentation.
- [] I updated the documentation where necessary.

### CODE REVIEW

**Comment by style95** [comment]
According to the order defined in the wiki, this component requires other components to be merged.
https://cwiki.apache.org/confluence/display/OPENWHISK/Component+Design

But there is no big difference in the memory queue logic, so I opened it for reviews.

Since this logic is the core logic and changes on this component are highly likely to introduce huge impacts on the system, I will write down a comprehensive design document.

Once dependent modules are merged into the master branch, I would rebase this PR.

File: common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * This buffer provides the average of the given elements.
+ * The number of elements are limited and the first element is removed if the maximum size is reached.
+ * Since it is based on the Vector, its operation takes effectively constant time.
+ * For more details, please visit https://docs.scala-lang.org/overviews/collections/performance-characteristics.html
+ *
+ * @param maxSize the maximum size of the buffer
+ */
+class AverageRingBuffer(private val maxSize: Int) {
```
Comment (style95) [inline common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala:15]: This circular buffer is used to calculate the average execution time of recent N activations for a given action.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/SchedulingDecisionMaker.scala [hunk 2]
```diff
@@ -0,0 +1,244 @@
+package org.apache.openwhisk.core.scheduler.queue
+
+import akka.actor.{Actor, ActorSystem, Props}
+import org.apache.openwhisk.common.Logging
+import org.apache.openwhisk.core.entity.FullyQualifiedEntityName
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+class SchedulingDecisionMaker(
```
Comment (style95) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/SchedulingDecisionMaker.scala:10]: This component decides whether to add more containers or not for the given action.
It uses the average execution time of an action.

There are three cases to calculate the average execution.
1. Initial execution.
In this case, we just assume the execution time as tick time(100ms by default).

2. Queue is newly created.
In this case, activations exist in the DB, but no data in the AverageRingBuffer.
So we fetch the average execution time from DB.

3. An action is continuously executed.
Each container sends the last execution time along with `FetchRequest` when it pulls a new activation from a queue.
And our scheduler keeps N execution times in the AverageRingBuffer. 
So we can easily calculate the average execution time of recent N executions.

File: common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package org.apache.openwhisk.common
```
Comment (ningyougang) [inline common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala:1]: Need to add licence header

**Comment by ningyougang** [comment]
It is better to add below configuration to scheduler's application.conf, e.g.
```
    queue {
      idle-grace = "20 seconds"
      stop-grace = "20 seconds"
      flush-grace = "60 seconds"
      graceful-shutdown-timeout = "5 seconds"
      max-retention-size = "10000"
      max-retention-ms = "60000"
      throttling-fraction = "0.9"
      duration-buffer-size = "10"
    }
```

Or when open another new pr which include ansible/roles/schedulers/tasks/deploy.yml, add there.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala [hunk 4]
```diff
@@ -0,0 +1,1096 @@
...
+    cleanUpActors(data)
+    cleanUpData()
+
+    context.parent ! queueRemovedMsg
+
+    goto(Removed) using NoData()
+  }
+
+  private def cleanUpActorsAndGotoRemovedIfPossible(data: RemovingData) = {
+    if (queue.isEmpty && requestBuffer.isEmpty) {
```
Comment (ningyougang) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:629]: should add `requestBuffer = requestBuffer.filter(!_.promise.isCompleted)` before this line: `if (queue.isEmpty && requestBuffer.isEmpty) {`

This is avoid `when updates queue,  in spite of all requests in requestBuffer will be finished finally, but the completed request may not be removed from requestBuffer`,so need to add `requestBuffer.filter(!_.promise.isCompleted)`

File: .travis.yml [hunk 5]
```diff
@@ -27,6 +27,7 @@ env:
   global:
     - ANSIBLE_CMD="ansible-playbook -i environments/local -e docker_image_prefix=testing"
     - GRADLE_PROJS_SKIP=""
+    - OW_SCALA_VERSION=2.13
```
Comment (ningyougang) [inline .travis.yml:30]: Great!

Comment (style95) [reply .travis.yml:30]: The scala compilation version was not the culprit.

**Comment by ningyougang** [comment]
Need rebase.

File: common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala (hunk 1 above)
Comment (bdoyle0182) [reply common/scala/src/main/scala/org/apache/openwhisk/common/AverageRingBuffer.scala:32]: What was the reasoning for picking average as the heuristic. Would median be a better heuristic here? i.e. if all activations take 100 milliseconds and then one activation has a slow call to a db that will heavily skew the average.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala [hunk 6]
```diff
@@ -0,0 +1,1097 @@
...
+
+    case Event(t: FailureMessage, _) =>
+      logging.error(this, s"[$invocationNamespace:$action:$stateName] got an unexpected failure message: $t")
+
+      stay
+
+    case Event(msg: DecisionResults, _) =>
+      val DecisionResults(result, num) = msg
+      result match {
+        case AddInitialContainer if num > 0 =>
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:550]: Should this be `if num = 0`

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala [hunk 7]
```diff
@@ -0,0 +1,1097 @@
...
+  }
+
+  private def handleActivationMessage(msg: ActivationMessage) = {
+    logging.info(this, s"[$invocationNamespace:$action:$stateName] got a new activation message ${msg.activationId}")(
+      msg.transid)
+    in += 1
+    takeUncompletedRequest()
+      .map { res =>
+        res.trySuccess(Right(msg))
+        in -= 1
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala:907]: Pretty sure this isn't thread safe if `takeUncompletedRequest` is a promise

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/SchedulingDecisionMaker.scala [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+          // do nothing
+          case _ =>
+            Future.successful(DecisionResults(Skip, 0))
+        }
+      }
+    }
+  }
+
+  private def addServersIfPossible(existing: Int,
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/SchedulingDecisionMaker.scala:206]: Should this be called `addContainersIfPossible`

**Comment by bdoyle0182** [comment]
This is probably the most complicated PR, but did the best I could to understand at a high level I think LGTM a part from my comments.

One comment on the persistence nature of the queue. I think queueing with kafka doesn't add anything to openwhisk if openwhisk does not provide at least once guarantee of activation execution. It's unnecessary overhead to queue up through kafka if the activation can get lost elsewhere anyways.

**Comment by style95** [comment]
> One comment on the persistence nature of the queue. I think queueing with kafka doesn't add anything to openwhisk if openwhisk does not provide at least once guarantee of activation execution. It's unnecessary overhead to queue up through kafka if the activation can get lost elsewhere anyways.

Agree.

I think the immediate next step could be removing Kafka from the critical path completely as currently, container creation messages and activation messages from controller to scheduler are delivered through Kafka.


> This queue is designed in a way that activations are not persistent.
But it could be improved in the future by taking a similar approach with what Kafka does.

I mentioned above that if we really need to support the at-least-once semantic, we can consider this approach.

**Comment by style95** [comment]
After rebasing this PR, it started failing to run compile `compileScoverageScala`.
It seems it works well with scala 2.13 but fails with 2.12.
Not quite sure the reason yet.

**Comment by style95** [comment]
It’s ready to merge.

**Pull request merged**